File: wear/src/main/java/com/atakmap/android/wickr/plugin/multisensortracking/ConnectionManager.java
/*
 * Copyright 2022 Samsung Electronics Co., Ltd. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.atakmap.android.wickr.plugin.multisensortracking;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;

import com.atakmap.android.wickr.plugin.R;
import com.samsung.android.service.health.tracking.ConnectionListener;
import com.samsung.android.service.health.tracking.HealthTracker;
import com.samsung.android.service.health.tracking.HealthTrackerException;
import com.samsung.android.service.health.tracking.HealthTrackingService;
import com.samsung.android.service.health.tracking.data.HealthTrackerType;

import java.util.List;

public class ConnectionManager {

    private final static String TAG = "Connection Manager";
    private final ConnectionObserver connectionObserver;
    private HealthTrackingService healthTrackingService = null;

    private final ConnectionListener connectionListener = new ConnectionListener() {
        @Override
        public void onConnectionSuccess() {
            Log.i(TAG, "Connected");
            connectionObserver.onConnectionResult(R.string.ConnectedToHs);
            if (!isSpO2Available(healthTrackingService)) {
                Log.i(TAG, "Device does not support SpO2 tracking");
                connectionObserver.onConnectionResult(R.string.NoSpo2Support);
            }
            if (!isHeartRateAvailable(healthTrackingService)) {
                Log.i(TAG, "Device does not support Heart Rate tracking");
                connectionObserver.onConnectionResult(R.string.NoHrSupport);
            }
        }

        @Override
        public void onConnectionEnded() {
            Log.i(TAG, "Disconnected");
        }

        @Override
        public void onConnectionFailed(HealthTrackerException e) {
            connectionObserver.onError(e);
        }
    };

    ConnectionManager(ConnectionObserver observer) {
        connectionObserver = observer;
    }

    public void connect(Context context) {
        healthTrackingService = new HealthTrackingService(connectionListener, context);
        healthTrackingService.connectService();
    }

    public void disconnect() {
        if (healthTrackingService != null)
            healthTrackingService.disconnectService();
    }

    public void initSpO2(SpO2Listener spO2Listener) {
        final HealthTracker healthTracker;
        healthTracker = healthTrackingService.getHealthTracker(HealthTrackerType.SPO2);
        spO2Listener.setHealthTracker(healthTracker);
        setHandlerForBaseListener(spO2Listener);
    }

    public void initHeartRate(HeartRateListener heartRateListener) {
        final HealthTracker healthTracker;
        healthTracker = healthTrackingService.getHealthTracker(HealthTrackerType.HEART_RATE);
        heartRateListener.setHealthTracker(healthTracker);
        setHandlerForBaseListener(heartRateListener);
    }

    private void setHandlerForBaseListener(BaseListener baseListener) {
        baseListener.setHandler(new Handler(Looper.getMainLooper()));
    }

    private boolean isSpO2Available(@NonNull HealthTrackingService healthTrackingService) {
        final List<HealthTrackerType> availableTrackers = healthTrackingService.getTrackingCapability().getSupportHealthTrackerTypes();
        return availableTrackers.contains(HealthTrackerType.SPO2);
    }

    private boolean isHeartRateAvailable(@NonNull HealthTrackingService healthTrackingService) {
        final List<HealthTrackerType> availableTrackers = healthTrackingService.getTrackingCapability().getSupportHealthTrackerTypes();
        return availableTrackers.contains(HealthTrackerType.HEART_RATE);
    }
}

File: wear/src/main/java/com/atakmap/android/wickr/plugin/multisensortracking/ConnectionManager.kt
package com.atakmap.android.wickr.plugin.multisensortracking

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.samsung.android.service.health.tracking.ConnectionListener
import com.samsung.android.service.health.tracking.HealthTrackerException
import com.samsung.android.service.health.tracking.HealthTrackingService
import com.samsung.android.service.health.tracking.data.HealthTrackerType

class ConnectionManager internal constructor(private val connectionObserver: ConnectionObserver) {

    interface ConnectionObserver {

        fun connected()

        fun disconnected()

        fun onError(healthTrackerException: HealthTrackerException)
    }

    var isConnected = false

    private var healthTrackingService: HealthTrackingService? = null
    private var availableTrackers = mutableListOf<HealthTrackerType>()

    private val connectionListener: ConnectionListener = object : ConnectionListener {
        override fun onConnectionSuccess() {
            isConnected = true
            connectionObserver.connected()
        }

        override fun onConnectionEnded() {
            isConnected = false
            connectionObserver.disconnected()
        }

        override fun onConnectionFailed(error: HealthTrackerException) {
            isConnected = false
            connectionObserver.onError(error)
        }
    }

    fun connect(context: Context?) {
        healthTrackingService = HealthTrackingService(connectionListener, context)
        healthTrackingService?.connectService()
        healthTrackingService?.trackingCapability?.supportHealthTrackerTypes?.let {
            availableTrackers = it
        }
    }

    fun disconnect() {
        healthTrackingService?.disconnectService()
    }

    fun setSpO2Listener(spO2Listener: SpO2Listener) {
        healthTrackingService?.getHealthTracker(HealthTrackerType.SPO2)?.let {
            spO2Listener.setHealthTracker(it)
            spO2Listener.setHandler(Handler(Looper.getMainLooper()))
        }
    }

    fun setHeartRateListener(heartRateListener: HeartRateListener) {
        healthTrackingService?.getHealthTracker(HealthTrackerType.HEART_RATE)?.let {
            heartRateListener.setHealthTracker(it)
            heartRateListener.setHandler(Handler(Looper.getMainLooper()))
        }
    }

    fun isSpO2Available(): Boolean {
        return availableTrackers.contains(HealthTrackerType.SPO2)
    }

    fun isHeartRateAvailable(): Boolean {
        return availableTrackers.contains(HealthTrackerType.HEART_RATE)
    }
}
